Remove requirements for hardware filters

cjdoucette opened the pull request:
Before these patches, Gatekeeper NICs were required to support EtherType and ntuple filters for hardware filters. These patches lift this requirement so that other NICs (such as the Amazon ENA) can still run Gatekeeper by filtering packets in software.

mengxiang0811 commented on lls/cache.h:
```diff
@@ -48,6 +48,18 @@ struct lls_put_req {
 
+/* Information needed to submit ARP packets to the LLS block. */
+struct lls_arp_req {
+	/* ND neighbor packets. */
```
Shouldn't it be ``/* ARP packets. */``?

mengxiang0811 commented on include/gatekeeper_net.h:
```diff
@@ -225,6 +225,9 @@ struct gatekeeper_if {
 	unsigned int       acl_func_count;
+
+	/* Whether the EtherType filter can be used on this interface. */
+	int                hw_filter_eth;
```
We can use a ``bool`` type here.

mengxiang0811 commented on lib/net.c:
```diff
@@ -116,19 +116,19 @@ ethertype_filter_add(uint8_t port_id, uint16_t ether_type, uint16_t queue_id)
-		RTE_LOG(ERR, PORT,
+		RTE_LOG(NOTICE, PORT,
 			"Port %hhu is invalid for adding an EtherType filter for 0x%02hx!\n",
 			ether_type, port_id);
```
This is a mistake from previous patch, we may switch the ``ether_type`` and ``port_id`` here.

mengxiang0811 commented on lls/cache.c:
```diff
@@ -311,6 +312,22 @@ lls_process_reqs(struct lls_config *lls_conf)
+				if (process_arp(lls_conf, arp->iface,
+						tx_queue, arp->pkts[i],
+						rte_pktmbuf_mtod(arp->pkts[i],
+						    struct ether_hdr *)) == -1)
```
Shouldn't we keep this line aligned with the above one?

cjdoucette replied:
Since there is a nested function call (to `rte_pktmbuf_mtod()`), I chose to indent its parameters to make it clearer that they are parameters for that function call, and not parameters for `process_arp()`.

cjdoucette commented on cps/main.c:
```diff
@@ -746,22 +818,45 @@ add_bgp_filters(struct gatekeeper_if *iface, uint16_t tcp_port_bgp,
+					iface->name);
+				return ret;
+			}
+		} else {
```
I don't think we can do that here, because we only want to execute the code in the `else` if the ntuple filter is supported. If we get rid of the `else`, then after successfully adding an IPv4 ACL in the `if`, we would also try to add an ntuple filter.

cjdoucette commented on lib/acl.c:
```diff
@@ -19,6 +19,308 @@
+			 */
+			for (j = 1; j < iface->acl4_func_count; j++) {
+				/* Skip over ACLs without matching function. */
+				if (iface->ext4_funcs[j] == NULL)
```
Not all of the ACL types have the possibility of an extension header. For example, we are assuming that packets destined for the GT-GK Unit are IPv4 and do not have an extension header. Therefore, when we add the ACL rule for the GGU packets, the argument for the `ext_cb_func` is `NULL`.

When we go to check all unrecognized packets for extension headers, we need to skip over the ACL types that do not have extension header functions, thus adding the `NULL` check here.

cjdoucette commented on gk/main.c:
```diff
@@ -1038,8 +1038,10 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 
-	process_pkts_ipv4_acl(&gk_conf->net->front, lcore, &acl4);
-	process_pkts_ipv6_acl(&gk_conf->net->front, lcore, &acl6);
+	process_pkts_acl(&gk_conf->net->front, lcore, &acl4,
```
I don't think so, because we need to know which ACL (IPv4 or IPv6) to use in each call of the function. If we just passed in the `struct gatekeeper_if`, we wouldn't know which ACL to use.

cjdoucette commented:
OK, the latest rebase has addressed all the concerns. I squashed the last two commits together, but I kept them separate on a different branch because I'm still unsure whether it's best to review them together or separately. But let me know what you think of these.

cjdoucette commented:
To get the KNI in the CPS block to work on EC2, I had to update the ENA driver in DPDK. Therefore, if you already have the gatekeeper repository, to see the changes to DPDK you'll need to run:

    git submodule update

And recompile DPDK and then recompile gatekeeper.

AltraMayor commented on include/gatekeeper_lls.h:
```diff
@@ -359,6 +361,10 @@ ipv6_addrs_equal(const uint8_t *addr1, const uint8_t *addr2)
 }
 
+/* Submit ARP packets to the LLS block (hardware filtering is not available). */
+int submit_arp(struct rte_mbuf **pkts, unsigned int num_pkts,
```
No caller of `submit_arp()` uses its return, and `submit_arp()` already deals with falures. Shouldn't we make the return of `submit_arp()` `void`?

cjdoucette commented:
Just rebased to incorporate the suggestion.

AltraMayor commented on ggu/main.c:
```diff
@@ -321,30 +325,98 @@ process_single_packet(struct rte_mbuf *pkt, const struct ggu_config *ggu_conf)
+	unsigned int i;
+	int ret;
+
+	RTE_VERIFY(num_pkts <= (sizeof(req->pkts) / sizeof(*req->pkts)));
```
Can't we use `RTE_DIM()` here?

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -203,31 +231,15 @@ struct gatekeeper_if {
 
 	/* Whether the EtherType filter can be used on this interface. */
-	bool               hw_filter_eth;
+	bool              hw_filter_eth;
```
This space adjustment should be done in the previous patch.

cjdoucette replied:
If it was done in the previous patch, it would appear to be one space off from all of the ACL fields that come before it.

Since this patch replaces those ACL fields with new fields with different spacing, this patch aligns this field to match up with the new fields.

cjdoucette commented on ggu/main.c:
```diff
@@ -400,11 +555,11 @@ run_ggu(struct net_config *net_conf,
 	}
 
-	ret = net_launch_at_stage1(net_conf, 0, 0, 1, 0, ggu_state1, ggu_conf);
+	ret = net_launch_at_stage1(net_conf, 0, 0, 1, 0, ggu_stage1, ggu_conf);
```
Previously many of the GT-GK Unit functions were named `ggu_state` instead of `ggu_stage`, so these changes are just replacing the `'t'` with a `'g'`.

cjdoucette commented:
Rebased with the changes.

AltraMayor commented on lls/main.c:
```diff
@@ -389,14 +409,19 @@ lls_proc(void *arg)
-				RTE_LOG(NOTICE, TIMER, "Can't reset front LACP timer to skip cycle\n");
+		int num_tx;
+
+		if (front->hw_filter_eth) {
```
Isn't it possible that a NIC does not support Ethernet filter and sends all non-IP packets to queue zero?

If so, the test in this `if` statement is not correct. And the same problem happens to the back interface.

AltraMayor commented on ggu/main.c:
```diff
@@ -423,9 +578,15 @@ run_ggu(struct net_config *net_conf,
-	ret = 0;
-	goto out;
+	ret = init_mailbox("ggu_mb", MAILBOX_MAX_ENTRIES,
+		sizeof(struct rte_mbuf *), ggu_conf->lcore_id,
```
Shouldn't we use `sizeof(struct ggu_request)` here instead of `sizeof(struct rte_mbuf *)`?

cjdoucette commented:
This rebase addressed the given suggestions.

AltraMayor commented on lib/acl.c:
```diff
@@ -73,6 +52,315 @@ drop_unmatched_ipv6_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
+	int ret;
+
+	switch (astate->proto) {
+	case ETHER_TYPE_IPv4:
```
This is the only place at which field `proto` of `struct acl_state` is used. And it has led to callers of `process_pkts_acl()` to write something like `&gt_conf->net->front.ipv4_acls` every time.

My solution to those problems:
1. Drop field `proto` of `struct acl_state`;
2. Replacing parameter `astate` of `process_pkts_acl()` with the protocol;
3. `process_pkts_acl()` should own this `switch...case` statement, and pass parameters`astate`, `proto_if_configured`, and `proto_name` to `process_acl()`.
 So passing the protocol instead of parameter `astate` would enable us to drop field `proto`. Moreover, this would simplify the call of `process_pkts_acl`, which now requires the caller to write something like `&gt_conf->net->front.ipv4_acls` every time.

AltraMayor replied:
Since `process_pkts_acl()` is inline and its parameter protocol is often going to be a constant, the compiler will eliminate the `switch...case` at compile time. A small bonus of the proposed solution.

cjdoucette commented:
The latest rebase addressed the suggestion.

AltraMayor commented on include/gatekeeper_acl.h:
```diff
@@ -79,12 +53,90 @@ add_pkt_ipv6_acl(struct acl_search *acl, struct rte_mbuf *pkt)
+		proto_name = "IPv4";
+		proto_if_configured = ipv4_if_configured(iface);
+		break;
+	case ETHER_TYPE_IPv6:
```
In a regular function, this code is fine. But given that this function is inline and we are trying to help the compiler to get rid of the `switch..case`, drop all current code in the cases `ETHER_TYPE_IPv4` and `ETHER_TYPE_IPv6`, and replace them with `return process_acl(...);`.

cjdoucette commented:
The latest rebase removes the `switch`.

cjdoucette commented:
The latest rebase *helps the compiler* remove the `switch`.

AltraMayor commented on lib/net.c:
```diff
@@ -185,14 +186,11 @@ ntuple_filter_add(uint8_t portid, uint32_t dst_ip,
+	if (!ipv4_configured)
+		goto ipv6;
 
 	if (dst_ip != 0) {
```
Thanks to the new `if (!ipv4_configured) goto ipv6;` above, we can drop this `if` test, can't we?

cjdoucette commented:
The rebase removed the unnecessary test.

AltraMayor commented on gt/main.c:
```diff
@@ -762,9 +765,13 @@ gt_proc(void *arg)
+				case ETHER_TYPE_IPv6:
 					add_pkt_ipv6_acl(&acl, m);
 					continue;
+				case ETHER_TYPE_ARP:
```
Somehow, this patch and the previous patch swap order. And this part of the code has a conflict.

cjdoucette commented:
I think it should be fixed now.
